### PR TITLE
Refactor lee, add sublayer-LEE and normalization, update exps

### PR DESCRIPTION
**Pull request by chanseok-lim**

### CODE REVIEW

File: exps_e2e.py [hunk 1]
```diff
@@ -14,62 +21,252 @@
...
+    print(f'MODEL_NAME: {MODEL_NAME}')
 
-    model = getattr(timm.models, args.modelname)(pretrained=True)
-    model.eval()
+    # load additionally available models
+    with open(CHKPT_LIST, 'r') as f:
+        chkpt_list = json.load(f)
 
-    evaluated_metrics = []
+    # load a model
```
Comment (chanseok-lim) [inline exps_e2e.py:188]: - Look up the model name in `chkpt_list`.  
  1. If found and `BACKBONE=True`, create a timm model using the specified backbone and `num_classes`, then load custom weights.  
  2. If found and `BACKBONE=False`, run the provided setup script (if any), then `torch.load` the serialized model file.  
- If not found, simply call `timm.create_model(MODEL_NAME, pretrained=True)`.  
- In all cases, `model.eval()` is called to switch to inference mode.

File: exps_e2e.py [hunk 2]
```diff
@@ -14,62 +21,252 @@
...
 
     metrics["dataset"] = key
-    metrics["model"] = args.modelname
+    metrics["model"] = model_name
     metrics["params"] = numparams(model)
 
     return metrics
 
+
+class Subnetwork(nn.Module):
```
Comment (chanseok-lim) [inline exps_e2e.py:68]: - Define `Subnetwork(nn.Module)` to wrap an existing model and extract the output of a specified intermediate layer (`top_layer`).
- In `__init__`:
  1. Store the original model and the target `top_layer` name.
  2. Assert that `top_layer` exists in `model.named_modules()`, raising an error if not.
- In `forward(x)`:
  1. Create an empty dictionary `outputs` to capture the hook output.
  2. Lookup the module object corresponding to `top_layer` from `self.original_model.named_modules()`.
  3. Define a `hook(module, input, output)` function that saves the layer’s output to `outputs[self.top_layer]`.
  4. Register the forward hook on the target module before running the model.
  5. Run `self.original_model(x)` to perform a full forward pass (the hook stores the intermediate output).
  6. Remove the hook immediately after the pass to restore the original model state.
  7. Return the captured feature tensor from `outputs[self.top_layer]`.

File: lee/e2e_lee.py
Comment (chanseok-lim) [inline lee/e2e_lee.py:1]: **Summary of Differences Between Old and New Versions in `lee/e2e_lee.py`**

---

### 1. Function Signature and Flexibility
- **Old Version**  
  ```python
  def get_equivariance_metrics(model, minibatch):
      # Only accepts a model and a minibatch.
      # Assumes a classifier model and fixed outputs.

- **New Version**
    ```python
    def get_equivariance_metrics(
        minibatch,
        model,
        model_type='classifier',
        output_type=torch.Tensor,
        output_key=None
    ):
        # - `model_type`: 'classifier', 'encoder', or 'subnetwork'
        # - `output_type`: type of output (e.g., torch.Tensor, dict)
        # - `output_key`: key/identifier to select the correct output from dict or tuple
    ```

- **Intention**
Make the function general-purpose so it can handle different architectures (classifiers, encoders, subnetworks) and various output formats.

---

### 2. Device Handling
- **Old Version**
  - Manually checks for CUDA availability.
  - Moves both the model and tensors to GPU if available (e.g., `model.cuda()`).

- **New Version**
  - Infers the device from the model’s parameters (`next(model.parameters()).device`).
  - Moves input tensors (e.g., `x.to(device)`) to that device.
  - Assumes the model is already on the correct device—no explicit `model.cuda()` call.

- **Intention**
Cleaner and more robust device handling that works across CPU/GPU setups, avoiding unnecessary model transfers.

---

### 3. Model Output Handling
- **Old Version**
  - Always applies `softmax` to the model’s raw output.
  - Assumes `model(...)` yields logits for classification.

- **New Version**
  - Branches logic depending on `model_type` and `output_type`.
  - Applies `softmax` only when appropriate—if `output_type` is a tensor of logits.
  - Provides error checking or fallback behavior for unexpected output types (e.g., dict or tuple).

- **Intention**
Increase compatibility with diverse models and output structures, making the function reusable beyond simple classifiers.

---

### 4. Lie Derivative Calculations
- **Old Version**
    ```python
    errs = {
        x: compute_<something>_lie_deriv(model_probs, x)
        for x in transformations
    }
    ```
    - Calls Lie‐derivative functions with just `model_probs` and `x`.

- **New Version**
    ```python
    errs = {
        x: compute_<something>_lie_deriv(model_probs, model_out, x)
        for x in transformations
    }
    ```
  - Also passes the precomputed `model_out` (raw output) to those functions.

- **Intention**
The derivative functions now require the full model output for correctness or efficiency—avoiding redundant computation inside those helper functions.

---

### 5. Metric Computation Logic
- **Old Version**
    ```python
    metrics = {
        x: pd.Series(errs[x].abs().cpu().data.numpy().mean(-1))
        for x in errs
    }
    ```
  - Directly computes the mean of the absolute error per metric.

- **New Version**
  - Introduces a helper function `compute_LEE` (or similar) that:
    1. Asserts the shape of the input tensor to ensure proper batch dimensions.
    2. Explicitly reshapes and averages over the intended axes.
  - Uses that helper to build `metrics` in a clearer, more robust way.

- **Intention**
Improve code clarity, ensure correct handling of batch dimensions, and make the metric‐computation step more maintainable.

### 6. General Code Improvements
- **Structure**
  - More modular breakdown—helper functions for metric computation and output‐handling logic.
  - Reduced code duplication and clearer control flow.

- **Error Handling**
  - Additional assertions and error messages (e.g., checking `output_key` exists if `output_type` is a dict).
  - Explicit `ValueError` or `AssertionError` when assumptions are violated.

- **Extensibility**
    - New parameters (`model_type`, `output_type`, `output_key`) allow future support for other architectures and output formats without rewriting the core function.
    - Easier to add support for new transformations or Lie‐derivative variants.

- **Overall Intention**
  - Generalize the function to support multiple neural‐network workflows (classifiers, encoders, subnetworks).
  - Improve robustness by handling devices, batch shapes, and output types more explicitly.
  - Enhance maintainability and readability, allowing future extension without major refactoring.
  - Avoid assumptions about model structure or output format, making the code compatible with a wider range of use cases.

File: lee/transforms.py
Comment (chanseok-lim) [inline lee/transforms.py:1]: Change for computing LEEs of subnetworks

File: lee/lie_derivs.py
Comment (chanseok-lim) [inline lee/lie_derivs.py:1]: Change for computing LEEs of subnetworks

**Comment by chanseok-lim** [comment]
Sorry, I accidentally created a pull request to the original repository instead of my fork!
